man/cards.conf.5.in


.PU
.TH cards.conf 5
.SH "NAME"
.LP 
cards.conf - Configuration file for cards

.SH "DESCRIPTION"
This file, by default
.IR /etc/cards.conf ,
is read by cards (see cards(8))
.LP
It might look like this:
.IP
.nf
###
# cards.conf

# root directories
dir /usr/ports/lxde
dir /usr/ports/graphic
dir /usr/ports/console
dir /usr/ports/base

dir /var/lib/pkg/remote/graphic-extra|http://downloads.nutyx.org
dir /var/lib/pkg/remote/graphic|http://downloads.nutyx.org
dir /var/lib/pkg/remote/console-extra|http://downloads.nutyx.org
dir /var/lib/pkg/remote/console|http://downloads.nutyx.org
dir /var/lib/pkg/remote/base-extra|http://downloads.nutyx.org
dir /var/lib/pkg/remote/base|http://downloads.nutyx.org


# write a log file

logdir /var/log/pkgbuild

# locales

locale en

# base system

base /usr/ports/base
.fi

.LP
There are a few values to be configured: the location where to look
for the ports, the location of the log file and the location of the 
base system packages.

.LP
Every line can contain one configuration option. The format of the lines
is
.B <key> <value>.
The valid keys are
.B dir
The value for 
.B prtdir 
is a directory, or a directory|an URL

.LP
The log options are
.B logfile
which represent the file path of the logs

.LP
The order of the dir options is important, as if a port is in multiple
directories, cards will use the one found first (directories listed
first have precedence)

.LP
If you specify an url in after a | symbol, it will used binaries from this url. In our example, we
will used the default mirror from nutyx.org.

.SH "AUTHORS"
Thierry Nuttens <tnut@nutyx.org>
.SH "SEE ALSO"
cards(8), pkgmk(8) pkgadd(8), ports(8), pkgmk.conf(5)


CHANGELOG.md
# v1.99.9 (xxxx-xx-xx)

## Summary

* Renamed to `p_cards` peter1000 development repo
* Updated: CHANGELOG.md, LICENSE.md, README.md, TODO.md
* Adds: initial documentation structure for [MkDocs](https://github.com/mkdocs/mkdocs)
* Removes french manpages
* Removes sub-package *fltk*
* Replaces `pkgmk.in` with `pkgmk2.in`
* Removes individual license headers in favour of the overall package license.
* Adds *Christophe Devine*, *Junio C Hamano* and *Logan Ingalls* to the Copyright holder in LICENSE.md

# History

* On 07. March 2016 **peter1000** <https://github.com/peter1000/> renamed 
    `cards commit 89580529259f5e912a2efc936bebf90e3f698537` to **p_cards**.

* [cards-1.99.8](https://github.com/NuTyX/cards/commit/992a2ca11bf771e0809ebe9d0987bdc1d47a8e82)    (on Mar 03, 2016)
* [cards-1.99.0](https://github.com/NuTyX/cards/commit/d9d4ae224175ab8514b1951a0078cc0fb75d60fc)    (on Jan 14, 2016)
* [cards-1.96.0](https://github.com/NuTyX/cards/commit/ae6c74e2f2f463bea0699ab2bb43f5daa3df15f2)    (on Jan 03, 2016)
* [cards-1.95.0](https://github.com/NuTyX/cards/commit/9f37977d754bd1dd88346f28e50ea629a69034d8)    (on Jan 02, 2016)
* [cards-1.4.0](https://github.com/NuTyX/cards/commit/f7bd5a8d341eb63aa0a772a5d8e3c201d2e97b15)     (on Dec 25, 2015)
* [cards-1.3.0](https://github.com/NuTyX/cards/commit/a695725e37beb64bda56ca000c5ee86d1f34541e)     (on Aug 25, 2015)
* [cards-1.2.1](https://github.com/NuTyX/cards/commit/14b580fc4165f256e952262914d9cb7e2cba6f74)     (on Jul 23, 2015)
* [cards-1.1](https://github.com/NuTyX/cards/commit/f3eb1243dabec33f7854dd6330a53438d62c8440)       (on Jun 06, 2015)
* [cards-1.0](https://github.com/NuTyX/cards/commit/49452594a9fc8c312fdc4c7203d76761ec5eab9f)       (on May 14, 2015) 


src/cards.cpp
#include <iostream>
#include <cstdlib>
#include "file_download.h"
#include "cards_base.h"
#include "cards_sync.h"
#include "cards_depends.h"
#include "cards_install.h"
#include "cards_remove.h"
#include "cards_create.h"
#include "cards_info.h"
#include "cards_upgrade.h"

#include "pkgrepo.h"
#include "pkginfo.h"
#include "pkgrm.h"

using namespace std;


int main(int argc, char** argv)
{
	string command = basename(argv[1]);
	string configFile = "/etc/cards.conf";
	try {
		CardsArgumentParser cardsArgPars;
		cardsArgPars.parse(argc, argv);

		if (cardsArgPars.isSet(CardsArgumentParser::OPT_CONFIG_FILE))
			configFile=cardsArgPars.getOptionValue(CardsArgumentParser::OPT_CONFIG_FILE);

		if (cardsArgPars.command() == CardsArgumentParser::CMD_HELP) {
			cout << "Usage: " << BLUE << cardsArgPars.appName()  << " command " << NORMAL << "[options]"<< endl;

			cout << "where possible " << BLUE << "command" << NORMAL<< " are:" << endl;

			cout << GREEN << "\nINFORMATION" << NORMAL << endl;
			cout << BLUE << "  help" << NORMAL << "                       show this help" << endl;
			cout << BLUE << "  list" << NORMAL << " [-b][-p]              show a list of installed packages" << endl
				<< "                             available binaries" << endl
				<< "                             available ports" << endl;

			cout << BLUE << "  info" << NORMAL << " [-b][-p] <package>    print info about install package" << endl
				<< "                             available binaries" << endl
				<< "                             available ports" << endl;
			cout << BLUE << "  config" << NORMAL << "                     show the configuration of your " <<  cardsArgPars.appName() << endl;
			cout << BLUE << "  level [-I]" << NORMAL << "                 show all the ports founds. The list is showned by order \
of compilation" << endl
				<< "                             If -I it will ignore the WARNING about NOT FOUND <dependencies> from <port>" << endl;
			cout << GREEN << "\nPORTS SPECIFIC SCENARIO" << NORMAL << endl;
			cout << BLUE << "  depends" << NORMAL << " [-i] <port>        show dependencies for the port in compilation order." << endl
				<< "                             If -i it will shows the installed dependencies as well." << endl;
			cout << BLUE << "  deptree" << NORMAL << "   <port>           show dependencies in a tree." << endl;
			cout << BLUE << "  depcreate" << NORMAL << " <port>           compile and install the port and its dependencies." << endl;
			cout << BLUE << "  create" << NORMAL << "    <port>           install all the dependencies from binaries and then compile the port." << endl;

			cout << GREEN << "\nDIFFERENCES / UPGRADE / CLEANUP" << NORMAL << endl;
			cout << BLUE << "  diff" << NORMAL << " [-p]                  list outdated packages." << endl
				<< "                             list outdated ports" << endl;
			cout << BLUE << "  upgrade" << NORMAL << "                    upgrade outdated packages." << endl;
			cout << BLUE << "  purge" << NORMAL << "                      cleanup downloaded binaries in cache." << endl;
			cout << GREEN << "\nSEARCHING" << NORMAL << endl;
			cout << BLUE << "  search" << NORMAL << " <expr>              show port names or description containing 'expr'" << endl;
			cout << BLUE << "  query" << NORMAL << "  <file>              list owner of file(s) matching the query." << endl;
			cout << BLUE << "  files" << NORMAL << "  <package>           list the file(s) owned by the <package>" << endl;
			cout << GREEN << "\nSYNCHRONISATION" << NORMAL << endl;
			cout << BLUE << "  sync" << NORMAL << "                       synchronize the local and remote meta datas." << endl;
			cout << GREEN << "\nINSTALL, UPDATE and REMOVAL" << NORMAL << endl;
			cout << BLUE << "  install" << NORMAL << " [-u][-f] <package> install the binary found on the mirror." << endl
				<< "                             If -u it will upgrade the installed package." << endl
				<< "                             If -f it will force the installation in case of files conflicts." << endl;
			cout << BLUE << "  remove" << NORMAL << " [-a] <package>      remove the installed package." << endl
				<< "                             If -a it will remove the sub-package as well." << endl;
			cout << GREEN << "\nBASE SYSTEM" << NORMAL << endl;
			cout << BLUE << "  base" << NORMAL << " -r                    return to a base system." << endl
				<< "                             You need to have a valid 'base' directory" << endl
				<< "                             define in /etc/cards.conf otherwise it will fails." << endl;
			return EXIT_SUCCESS;
		}
		if (cardsArgPars.command() == CardsArgumentParser::CMD_CONFIG) {
			Config config;
			Pkgrepo::parseConfig(configFile, config);
			unsigned int index = 0;
			string prtDir, url ;
			for (vector<DirUrl>::iterator i = config.dirUrl.begin();i != config.dirUrl.end();++i) {
				index++;
				DirUrl DU = *i ;
				prtDir = DU.Dir;
				url = DU.Url;
				cout << index << " Directory: " << prtDir ;
				if ( url != "" ) {
					cout << " from " << url  << endl;
				} else {
					cout << endl;
				}
			}
			for ( auto i : config.baseDir ) cout << "Base System list directory: " << i << endl;
			cout <<   "Binaries : " << config.arch << endl;
			for ( auto i : config.locale ) cout << "Locale   : " << i << endl;
			if ( config.logdir != "") {
				cout << "log directory: " << config.logdir << endl;
			}
			return EXIT_SUCCESS;
		} else if (cardsArgPars.command() == CardsArgumentParser::CMD_DEPCREATE) {
			// get the list of the dependencies
			CardsDepends CD(cardsArgPars);
			vector<string> listOfPackages = CD.getNeededDependencies();
			if ( listOfPackages.empty() ) {
				cout << "The package " << cardsArgPars.otherArguments()[0] << " is already installed" << endl;
				return EXIT_SUCCESS;
			}
			// create (compile and install) the List of deps (including the final package)
			Cards_create CC(cardsArgPars,configFile,listOfPackages);
			return EXIT_SUCCESS;
		} else if (cardsArgPars.command() == CardsArgumentParser::CMD_CREATE) {

			if ( ( ! cardsArgPars.isSet(CardsArgumentParser::OPT_REMOVE)) &&
			( ! cardsArgPars.isSet(CardsArgumentParser::OPT_DRY)) ) {
				cardsArgPars.printHelp("create");
				return EXIT_SUCCESS;
			}
			if  ( ! cardsArgPars.isSet(CardsArgumentParser::OPT_DRY)) {
				// go back to a base system
				Cards_base CB(cardsArgPars);
				CB.run(argc, argv);
			}
			// get the list of the dependencies"
			CardsDepends CD(cardsArgPars);
			vector<string> listOfDeps = CD.getDependencies();

			if (!listOfDeps.empty())
				Cards_install CI(cardsArgPars,configFile,listOfDeps);

			// compilation of the final port"
			Cards_create CC( cardsArgPars,configFile,cardsArgPars.otherArguments()[0]);
			return EXIT_SUCCESS;

		} else if (cardsArgPars.command() == CardsArgumentParser::CMD_BASE) {
			if ( ( ! cardsArgPars.isSet(CardsArgumentParser::OPT_REMOVE)) &&
			( ! cardsArgPars.isSet(CardsArgumentParser::OPT_DRY)) ) {
				cardsArgPars.printHelp("base");
				return EXIT_SUCCESS;
			}
			Cards_base CB(cardsArgPars);
			CB.run(argc, argv);
			return EXIT_SUCCESS;
		} else if (cardsArgPars.command() == CardsArgumentParser::CMD_SYNC) {
			Cards_sync CS(cardsArgPars);
			CS.run();
			return EXIT_SUCCESS;
		} else if (cardsArgPars.command() == CardsArgumentParser::CMD_PURGE) {
			Cards_sync CS(cardsArgPars);
			CS.purge();
			return EXIT_SUCCESS;
		} else if ( (cardsArgPars.command() == CardsArgumentParser::CMD_UPGRADE) ||
					(cardsArgPars.command() == CardsArgumentParser::CMD_DIFF) ) {
			Cards_upgrade CU(cardsArgPars,configFile);
			return EXIT_SUCCESS;
		} else if (cardsArgPars.command() == CardsArgumentParser::CMD_INSTALL) {
				Cards_install CI(cardsArgPars,configFile);
			return EXIT_SUCCESS;
		} else if (cardsArgPars.command() == CardsArgumentParser::CMD_REMOVE) {
			Cards_remove CR("cards remove",cardsArgPars, configFile);
			return EXIT_SUCCESS;
		} else if (cardsArgPars.command() == CardsArgumentParser::CMD_LEVEL) {
			CardsDepends CD(cardsArgPars);
			CD.showLevel();
			return EXIT_SUCCESS;
		} else if (cardsArgPars.command() == CardsArgumentParser::CMD_DEPENDS) {
			CardsDepends CD(cardsArgPars);
			CD.showDependencies();
			return EXIT_SUCCESS;
		} else if (cardsArgPars.command() == CardsArgumentParser::CMD_DEPTREE) {
			CardsDepends CD(cardsArgPars);
			return CD.deptree();
		} else if ( (cardsArgPars.command() == CardsArgumentParser::CMD_INFO) ||
				(cardsArgPars.command() == CardsArgumentParser::CMD_LIST)   ||
				(cardsArgPars.command() == CardsArgumentParser::CMD_SEARCH) ||
				(cardsArgPars.command() == CardsArgumentParser::CMD_FILES)  ||
				(cardsArgPars.command() == CardsArgumentParser::CMD_QUERY)) {
			Cards_info CI(cardsArgPars,configFile);
			return EXIT_SUCCESS;
		}
	} catch (runtime_error& e) {
			cerr << "cards " << VERSION << " "<< command << ": " << e.what() << endl;
			return EXIT_FAILURE;
	}
}
// vim:set ts=2 :
